feat: add support for targetting specific workspaces

MalickBurger opened the pull request:
Implements #1126

jkowalleck commented on CONTRIBUTING.md:
````diff
@@ -47,5 +47,5 @@ Please sign off your commits, to show that you agree to publish your changes und
 , and to indicate agreement with [Developer Certificate of Origin (DCO)](https://developercertificate.org/).
 
 ```shell
-git commit --signed-off ...
````
this change appear to be unrelated to the intended feature.
please revert it here and open a dedicated PR for it.

jkowalleck replied:
fixed via 5af703ae20cd7ce9ff1b3d11b8ed9bea82e71c09

jkowalleck commented on src/builders.ts:
```diff
@@ -166,6 +169,10 @@ export class BomBuilder {
     }
 
+    for (const workspace of this.workspaces) {
+      args.push(`--workspace=${workspace}`)
```
for better intercompatibility with npm, please use 
```javascript
args.push('--workspace', workspace)
```

jkowalleck commented on src/cli.ts:
```diff
@@ -166,6 +167,12 @@ function makeCommand (process: NodeJS.Process): Command {
       Enums.ComponentType.Application
     )
+  ).addOption(
+    new Option(
```
for the things we pass trugh to npm, it is intended to mimic the CLI of npm.
the proposed option `workspaces` here does not do so.

compare with npm ls:
```text
List installed packages

Usage:
npm ls <package-spec>

Options:
[-a|--all] [--json] [-l|--long] [-p|--parseable] [-g|--global] [--depth <depth>]
[--omit <dev|optional|peer> [--omit <dev|optional|peer> ...]] [--link]
[--package-lock-only] [--no-unicode]
[-w|--workspace <workspace-name> [-w|--workspace <workspace-name> ...]]
[-ws|--workspaces] [--include-workspace-root] [--install-links]

alias: list

Run "npm help ls" for more info
```

jkowalleck commented on src/builders.ts:
```diff
@@ -166,6 +169,10 @@ export class BomBuilder {
 
+    for (const workspace of this.workspaces) {
+      args.push(`--workspace=${workspace}`)
+    }
```
when enabling workspaces, we would also need an option to enable workspace root.

MalickBurger replied:
I'm not sure I 100% agree/understand the point of this flag from `npm ls`. If I understand correctly, the current behaviour of this plugin when run in a root project is that all dependencies (which includes all workspace dependencies) are included in the output.

The feature requested was to be able to provide a specific workspace when running the plugin in the root so that only dependencies for that specific workspace is included in the the output.

The docs describes the workspace root option as:

```
[include-workspace-root](https://docs.npmjs.com/cli/v10/commands/npm-ls#include-workspace-root)
Default: false
Type: Boolean
Include the workspace root when workspaces are enabled for a command.

When false, specifying individual workspaces via the workspace config, or all workspaces via the workspaces flag, will cause npm to operate only on the specified workspaces, and not on the root project.
```

Does this not mean that if we set this property to true we go back to the original logic (no workspaces provided, use all dependencies found in root project) and if it is set to false (default) then we get our expected output meaning we don't need it?

This is also valid for the `workspaces` command which applies the logic to all workspaces (effectively the same as specifying no workspaces). I'm not sure if these options are not just copied over from other cli commands (such as `install`) and don't make sense for `ls` or if I am misunderstanding something.

MalickBurger replied:
Hi @jkowalleck, any update regarding my comments above?

jkowalleck replied:
need to experiment with `npm ls` a bit more. 
added additional demo/setups via https://github.com/CycloneDX/cyclonedx-node-npm/pull/1215 to do so 

will come back later, when i have a deeper insight.

jkowalleck replied:
> when enabling workspaces, we would also need an option to enable workspace root.

after experimenting with `npm ls --include-workspace-root -w ...`, if found it not having any effect on the results.
 so we do not need it.

all good.

jkowalleck reviewed (changes requested):
some things require chen. see my remarks.

also, please add proper unit tests

MalickBurger commented:
> some things require chen. see my remarks.
> 
> also, please add proper unit tests

Happy to do the changes. I would think some integration tests would be required for this change as it is not really feasible to test it in isolation due to the logic effectively sitting on the `npm ls` side.

I agree with the requirement of testing. I spent some time initially before making the PR going through the existing unit and integration tests and it is not very clear to me how we should go about adding tests for new functionality. For example, I do not see tests for several of the CLI options (such as `--short-PURLs`).

Could you provide a bit of a guide regarding how we should lay out tests for new functionality going forward as well as a brief of how the existing integration tests work and how they are structured. I intend on making additional PRs in the future for other issues and features so this would be very beneficial for me (and I'm sure other community members).

jkowalleck commented:
@MalickBurger for integration tests, 
1. you would add new setups here, if needed: <https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/demo>
2. Then, add the appropriate `npm ls ...` run here: <https://github.com/CycloneDX/cyclonedx-node-npm/blob/main/.github/workflows/npm-ls_demo-results.yml>
3. have the `npm ls ...` result data gathered via a CI (Github action/workflow). 
   you can dispatch the action in your repo here: <https://github.com/MalickBurger/cyclonedx-node-npm/actions> by dispatching the workflow (you might need to enable workflows for your repo)
4. download the action artifacts and add them here: <https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/tests/_data/npm-ls_demo-results>
5. remove unnecessary data via <https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/tests/_data/npm-ls_demo-results/_duplicates-cleaner> (might require a back-merge or rebase to pull the scripts in)

since you are planning on adding conditional parameters in the `npm ls` call, you need to incorporate this in the existing testing processes somehow.

MalickBurger commented:
> @MalickBurger, for adding additional integration tests,
> 
> 1. you would add new setups here, if needed: https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/demo
> 2. Then, add the appropriate `npm ls ...` run here: https://github.com/CycloneDX/cyclonedx-node-npm/blob/main/.github/workflows/npm-ls_demo-results.yml
> 3. have the `npm ls ...` result data gathered via a CI (Github action/workflow).
>    you can dispatch the action in your repo here: https://github.com/MalickBurger/cyclonedx-node-npm/actions by dispatching the workflow (you might need to enable workflows for your repo)
> 4. download the action artifacts and add them here: https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/tests/_data/npm-ls_demo-results
> 5. remove unnecessary data via https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/tests/_data/npm-ls_demo-results/_duplicates-cleaner (might require a back-merge or rebase to pull the scripts in)
> 
> since you are planning on adding conditional parameters in the `npm ls` call, you need to incorporate this in the existing testing processes somehow.

Looking at this, could you provide some testing support for an existing argument that is passed to `npm ls` by the plugin which I could then base my additional tests on? It would be easier and cleaner for a code maintainer to set these up so there is a standard to follow going forward for community members.

jkowalleck commented:
> > @MalickBurger, for adding additional integration tests,
> > 
> > 1. you would add new setups here, if needed: https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/demo
> > 2. Then, add the appropriate `npm ls ...` run here: https://github.com/CycloneDX/cyclonedx-node-npm/blob/main/.github/workflows/npm-ls_demo-results.yml
> > 3. have the `npm ls ...` result data gathered via a CI (Github action/workflow).
> >    you can dispatch the action in your repo here: https://github.com/MalickBurger/cyclonedx-node-npm/actions by dispatching the workflow (you might need to enable workflows for your repo)
> > 4. download the action artifacts and add them here: https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/tests/_data/npm-ls_demo-results
> > 5. remove unnecessary data via https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/tests/_data/npm-ls_demo-results/_duplicates-cleaner (might require a back-merge or rebase to pull the scripts in)
> > 
> > since you are planning on adding conditional parameters in the `npm ls` call, you need to incorporate this in the existing testing processes somehow.
> 
> Looking at this, could you provide some testing support for an existing argument that is passed to `npm ls` by the plugin which I could then base my additional tests on? It would be easier and cleaner for a code maintainer to set these up so there is a standard to follow going forward for community members.

will do. just bare with me, it may take a while

jkowalleck commented:
> will do. just bare with me, it may take a while

I am currently working on a solution. Will ping as soon as I have something for you.

jkowalleck commented on src/cli.ts:
```diff
@@ -166,6 +167,12 @@ function makeCommand (process: NodeJS.Process): Command {
     )
+  ).addOption(
+    new Option(
+      '--workspaces <workspace...>',
```
```suggestion
      '-w, --workspace <workspace-name...>',
```

jkowalleck replied:
An option `'-ws, --workspaces'` is to be added, too. it should be piped to `npm-ls`, just like `'-w, --workspace ...'`

jkowalleck commented:
@MalickBurger , 

I've overhauled the integration tests.
It would currently be enough to add your tests, showcasing that the intended parameters are sent to `npm-ls`. This could be done here: <https://github.com/CycloneDX/cyclonedx-node-npm/blob/main/tests/integration/cli.args-pass-through.test.js>

you might add something like 
```js
// region workspace
['workspace not supported npm 6 ', `6.${rMinor}.${rPatch}`, ['--workspace my-ws'], [...npm6ArgsGeneral]],
['package-lock-only npm 7', `7.${rMinor}.${rPatch}`, ['--workspace my-ws'], [...npm7ArgsGeneral, '--workspace my-ws']],
['package-lock-only npm 8', `8.${rMinor}.${rPatch}`, ['--workspace my-ws'], [...npm8ArgsGeneral, '--workspace my-ws']],
['package-lock-only npm 9', `9.${rMinor}.${rPatch}`, ['--workspace my-ws'], [...npm9ArgsGeneral, '--workspace my-ws']],
['package-lock-only npm 10', `10.${rMinor}.${rPatch}`, ['--workspace my-ws'], [...npm10ArgsGeneral, '--workspace my-ws']]
// endregion workspace
```

jkowalleck commented on src/cli.ts:
```diff
@@ -173,6 +176,22 @@ function makeCommand (process: NodeJS.Process): Command {
+    ).default([], 'empty')
+  ).addOption(
+    new Option(
+      '-no-ws, --no-workspaces',
```
why is this needed? 
which version of `npm ls` knowns this?

jkowalleck replied:
i understand. good catch.

jkowalleck commented on src/cli.ts:
```diff
@@ -173,6 +176,22 @@ function makeCommand (process: NodeJS.Process): Command {
+    )
+  ).addOption(
+    new Option(
+      '-wr, --include-workspace-root',
```
where does the `-wr`  come from? 

it is intended to use similar args like `nom ls` does, so adoption is easy.
which version of npm knows `-wr` here?

MalickBurger replied:
There is no shorthand from `npm ls`, assumed we would add our own if there isn't any defined, happy to remove

jkowalleck replied:
i see. no worries. 
We dont add custom shorthands. we try to mimic `npm ls` as much as possible, so people that know `npm ls` are not surprised.

jkowalleck commented on tests/integration/cli.args-pass-through.test.js:
```diff
@@ -72,6 +72,30 @@ describe('integration.cli.args-pass-through', () => {
+      // endregion workspace
+      // region include-workspace-root
+      ['workspace root not supported npm 6', `6.${rMinor}.${rPatch}`, ['-w', 'my-wsA', '--include-workspace-root'], [...npm6ArgsGeneral]],
+      ['workspace root npm 7', `7.${rMinor}.${rPatch}`, ['--workspace', 'my-wsA', '-w', 'my-wsB', '--include-workspace-root'], [...npm7ArgsGeneral, '--workspace=my-wsA', '--workspace=my-wsB', '--include-workspace-root=true']],
```
workspace root not supported npm 7

jkowalleck commented on src/cli.ts:
```diff
@@ -52,6 +52,9 @@ interface CommandOptions {
   outputFile: string
   validate: boolean | undefined
   mcType: Enums.ComponentType
+  workspace: string[]
```
please move the worksapce-related options up so they appear right below the `omit`

jkowalleck commented on src/cli.ts:
```diff
@@ -173,6 +176,22 @@ function makeCommand (process: NodeJS.Process): Command {
     ).default(
       Enums.ComponentType.Application
     )
+  ).addOption(
```
please move the worksapce-related options up so they appear right below the `omit`

jkowalleck commented:
merged in the latest master, fixed some merge conflicts.
did another round of review. result: there are open questions and issues

MalickBurger commented:
> merged in the latest master, fixed some merge conflicts. did another round of review. result: there are open questions and issues

Will take a look and update over the next few days when I have a chance

jkowalleck commented:
seams like some tests are failing.

MalickBurger commented:
> seams like some tests are failing.

Should be resolved now.

jkowalleck commented:
i am considdering this feature to be integrated as "experimental".

MalickBurger commented:
Hi @jkowalleck just want to find out where we stand regarding this PR. From my side I believe it is ready to merge.

jkowalleck commented on src/builders.ts:
```diff
@@ -172,6 +181,32 @@ export class BomBuilder {
+      }
+    }
+
+    // No need to set explicitly if true as this is default behaviour
```
```suggestion
    // No need to set explicitly if true as this seams to be default behaviour for `npm-ls`
```

jkowalleck commented on tests/integration/cli.args-pass-through.test.js:
```diff
@@ -70,8 +70,32 @@ describe('integration.cli.args-pass-through', () => {
+      ['workspaces disabled npm 8', `8.${rMinor}.${rPatch}`, ['--no-workspaces'], [...npm8ArgsGeneral, '--workspaces=false']],
+      ['workspaces disabled npm 9', `9.${rMinor}.${rPatch}`, ['--no-workspaces'], [...npm9ArgsGeneral, '--workspaces=false']],
+      ['workspaces disabled npm 10', `10.${rMinor}.${rPatch}`, ['--no-workspaces'], [...npm10ArgsGeneral, '--workspaces=false']],
+      ['workspaces disabled npm 11', `11.${rMinor}.${rPatch}`, ['--no-workspaces'], [...npm11ArgsGeneral, '--workspaces=false']]
```
this flag seams unknown to `npm`
```sellSession
$ npm11 ls --help
List installed packages

Usage:
npm ls <package-spec>

Options:
[-a|--all] [--json] [-l|--long] [-p|--parseable] [-g|--global] [--depth <depth>]
[--omit <dev|optional|peer> [--omit <dev|optional|peer> ...]]
[--include <prod|dev|optional|peer> [--include <prod|dev|optional|peer> ...]]
[--link] [--package-lock-only] [--no-unicode]
[-w|--workspace <workspace-name> [-w|--workspace <workspace-name> ...]]
[-ws|--workspaces] [--include-workspace-root] [--install-links]

alias: list

Run "npm help ls" for more info
```

jkowalleck replied:
though it is not documented, it still has effect.
it is the same as `npm ls --workspaces=false`

MalickBurger replied:
> Another reason I did it this way is because of how poor the documentation for `npm ls` is for this specific argument, this is what the official docs say:
> 
> ```
> [workspaces](https://docs.npmjs.com/cli/v11/commands/npm-ls#workspaces)
> Default: null
> Type: null or Boolean
> Set to true to run the command in the context of all configured workspaces.
> 
> Explicitly setting this to false will cause commands like install to ignore workspaces altogether. When not set explicitly:
> 
> Commands that operate on the node_modules tree (install, update, etc.) will link workspaces into the node_modules folder. - Commands that do other things (test, exec, publish, etc.) will operate on the root project, unless one or more workspaces are specified in the workspace config.
> ```

How does a user determine what the default behaviour is from this documentation.

jkowalleck replied:
I see.

My suggestion: We use `--workspaces` and have it a boolean with default value of `null`. (like npm docssay)
if the value of `options.workspaces` is not `null`, then the value is passed to `npm-ls`.
And of course, our own help text should be meaningful. Unlike the one from npm-ls


What do you think?

MalickBurger replied:
I like it, I think its a good compromise, I will make the relevant updates

jkowalleck commented on src/cli.ts:
```diff
@@ -87,6 +90,25 @@ function makeCommand (process: NodeJS.Process): Command {
+    ).default([], 'empty')
+  ).addOption(
+    new Option(
+      '--no-workspaces',
```
isthis option even used?

MalickBurger replied:
Yes, it sets the `--workspaces` option to `false`, this is a standard approach when using Commander.

See: https://github.com/tj/commander.js/blob/HEAD/examples/options-negatable.js and https://www.npmjs.com/package/commander

jkowalleck reviewed (changes requested):
something seams not right

jkowalleck commented on src/cli.ts:
```diff
@@ -87,6 +90,25 @@ function makeCommand (process: NodeJS.Process): Command {
+    ).default([], 'empty')
+  ).addOption(
+    new Option(
+      '--no-workspaces',
```
i still do not see how this comes together.
where comes this `--no-workspaces` from? I mean, npm knows `-ws|--workspaces`  -and it defaults to `null` for this. and you can set tot to true/false ... 
why does this option not mimic npm-ls options?

MalickBurger replied:
> i still do not see how this comes together. where comes this `--no-workspaces` from? I mean, npm knows `-ws|--workspaces` -and it defaults to `null` for this. and you can set tot to true/false ... why does this option not mimic npm-ls options?

Npm ls does support it, test it out.

MalickBurger replied:
> i still do not see how this comes together. where comes this `--no-workspaces` from? I mean, npm knows `-ws|--workspaces` -and it defaults to `null` for this. and you can set tot to true/false ... why does this option not mimic npm-ls options?

I don't follow you, this is best commander.js supports booleans unless we parse the value as a string and convert to a boolean https://github.com/tj/commander.js/issues/344.

MalickBurger replied:
> please remove this option. we tend to mimic `npm-ls` - and they dont call this out explicitely.

They also don't explicitly mention how workspace logic works by default but we are now telling users how that works through or params and README (which we have to do because it is not intuitive at all so they need some guidance)

MalickBurger replied:
> > please remove this option. we tend to mimic `npm-ls` - and they dont call this out explicitely.
> 
> They also don't explicitly mention how workspace logic works by default but we are now telling users how that works through or params and README (which we have to do because it is not intuitive at all so they need some guidance)

Additionally, without it, how will users disable the logic?

jkowalleck reviewed (changes requested):
i think some parts need discussion

MalickBurger commented on src/builders.ts:
```diff
@@ -172,6 +181,31 @@ export class BomBuilder {
+    // No need to set explicitly if false as this is default behaviour
+    if (this.includeWorkspaceRoot) {
+      if (npmVersionT[0] >= 8) {
+        args.push('--include-workspace-root=true')
```
No, this can't be set to false it is a Commander boolean and we have no "--no-include-workspace-root" defined so if it is defined it is true..

MalickBurger commented on src/builders.ts:
```diff
@@ -64,6 +67,9 @@ export class BomBuilder {
   gatherLicenseTexts: boolean
+  workspace: string[]
+  includeWorkspaceRoot: boolean
+  workspaces?: boolean
```
From the builder perspective we can treat this as undefined as well and thus don't pass to npm ls (default behaviour) but will change it if you prefer.

MalickBurger commented on src/builders.ts:
```diff
@@ -172,6 +181,31 @@ export class BomBuilder {
+      }
+    }
+
+    // No need to set explicitly if false as this is default behaviour
```
That would mean we are changing default behaviour and it is not an opt-in experimental feature, up until now this was not explicitly provided as false and it will be going forward. If you are happy with that then I can remove this check.

jkowalleck replied:
i see. interesting point. i think you are right. 
give me some time to let this sink in.

jkowalleck replied:
clarified

MalickBurger commented on src/builders.ts:
```diff
@@ -86,6 +92,9 @@ export class BomBuilder {
     this.shortPURLs = options.shortPURLs ?? false
     this.gatherLicenseTexts = options.gatherLicenseTexts ?? false
+    this.workspace = options.workspace ?? []
+    this.includeWorkspaceRoot = options.includeWorkspaceRoot ?? false
```
I don't think It is necessary to introduce another boolean with a third value of logic (undefined), it is disabled by npm by default and must be explicitly enabled which we are following here. If npm changes the default logic in the future then we will either need to update this here or we will need to document that there is different behaviour for different npm versions. Both options would require changes.

jkowalleck replied:
agree

MalickBurger commented on src/cli.ts:
```diff
@@ -238,6 +270,20 @@ export async function run (process: NodeJS.Process): Promise<number> {
   }
 
+  if (options.workspaces === false) {
+    if (options.workspace !== undefined && options.workspace.length > 0) {
```
Yes, this is remnant of previous logic, will remove the check

MalickBurger commented on src/cli.ts:
```diff
@@ -238,6 +270,20 @@ export async function run (process: NodeJS.Process): Promise<number> {
+  if (options.includeWorkspaceRoot) {
+    if (options.workspace.length === 0 && options.workspaces !== true) {
+      myConsole.error('ERROR | Bad config: `--include-workspace-root` can only be used when `--workspace` or `--workspaces` is also configured')
+      throw new Error('bad config')
```
Well it depends on how you look at it, the "--include-workspace-root" param only has an impact or relevance if you are specifying "--workspace=x" or "--workspaces" because all other scenarios we are not dealing with workspaces explicitly so why use the param? 

In this case the logic is slightly wrong because I am not handling the default behaviour of "workspaces" which is effectively true (but is actually null.........).

MalickBurger replied:
The reason I included this was to help guide users a bit on how it all works together, so they don't do things like always pass in "--include-workspace-root" thinking they need to even though root dependencies are always included unless under certain circumstances (which I document in the error and README/cli).

MalickBurger commented:
I think we are conflicting slightly on how this should be implemented because I am trying to design this in a way that is the most intuitive and yet still flexible experience for users. The aim is for it to be clear what the default behaviour is and how they can modify this behaviour if they wish.

The way you would like this done is to mimic npm ls exactly which I still don't fully understand since most users of the plugin wouldn't really care how the internals work they just want a tool that can create sboms for npm projects as easily as possible.

In an ideal world we could both have our way but I don't think it is the case here because the behaviour of npm ls with regards to workspaces is non intuitive and a bit of a black box (it also changes between version 7 and newer versions).

Before I go back and rewrite the logic again let's please fully decide how this is going to look, how we are going to explain it to users, and what default logic will be for all combination of workspace related parameters.

jkowalleck commented on src/cli.ts:
```diff
@@ -43,6 +43,9 @@ interface CommandOptions {
   omit: Omittable[]
+  workspace: string[]
+  includeWorkspaceRoot: boolean
+  workspaces: boolean | undefined
```
agree

jkowalleck commented:
> The way you would like this done is to mimic npm ls exactly which I still don't fully understand since most users of the plugin wouldn't really care how the internals work they just want a tool that can create sboms for npm projects as easily as possible.

This PR and a reading `npm` docs and changelogs showed ma, that even `npm` developer had no clear vision how things should be working. the idea of workspaces might look simple at first, but it has so many "use cases", more than they imagined at first.

I had the idea to mimic `npm`, because i thought users knew how their `npm` worked, but were just not sure how to generate SBOM from the data they had.
If we mimicked `npm`, then we would have not to take care of all the strange edge cases every other npm version introduced - we would lay this into the hand of the users.

My idea was to go the "easy" path - from a product owner's perspective - as all responsibility would be shifted to the users. Especially with workspaces, I would expect to have experienced users who knew what they do.
Our discussions showed me, that my idea was bad from a user's perspective. They now would need to learn about their own npm and its edge cases. Now they would need to understand aspects they maybe did not even care about. This could be considered as bad user experience.

I think I understand your point, and you convinced me.
If your idea was to produce consistent results over all supported npm versions, then the current tests might be insufficient. For this reason, I would say: 
**implement it the way you envisioned it. :+1:, and if needed, get rid of current tests** 

After the implementation is done, we might see which kind of test data is required and how it can be provided.
Since the feature is experimental, i'd suggest adding test data in a separate pullrequest later.

mapau commented:
@MalickBurger and @jkowalleck what is the status here? I would like to use the cyclonedx-node-npm with some projects that uses npm workspaces. Let me know if I can help here in any way. I will try, the branch build on my projects for now.

MalickBurger commented:
> @MalickBurger and @jkowalleck what is the status here? I would like to use the cyclonedx-node-npm with some projects that uses npm workspaces. Let me know if I can help here in any way. I will try, the branch build on my projects for now.

Hi @mapau, I will be getting back to this later in the week, we are close (hopefully) to having this PR finalized. Happy if you would consider doing some testing once complete?

MalickBurger commented:
Hi @jkowalleck,

I have refactored my changes based on our last discussion. I have simplified the logic and have tried to make it as simple and intuitive as possible to users while still mimicking npm ls where possible. I have designed it in such a way that users do not need to have an indepth understanding of npm ls or knowledge on which version they are running.

I have updated the tests but I need to do a few manual sanity tests as well however I will wait for now incase I need to make any additional changes.

jkowalleck reviewed (changes requested):
i have only implementation-related remarks.

the UX looks good. 
what do you think, @mapau?

mapau commented:
@jkowalleck I used the branch build on my projects that used npm workspace. All works as expected. The CLI help and info is clear and I would use this version as it is.

Only small thing, on my zsh the help output from the --no-workspaces section is not aligned with the other, see screenshot.

<img width="735" alt="Screenshot 2025-03-17 at 13 26 11" src="https://github.com/user-attachments/assets/b58329f4-a3d9-4ad7-86f0-36f67ccbe6f6" />

MalickBurger commented on src/cli.ts:
```diff
@@ -238,6 +272,23 @@ export async function run (process: NodeJS.Process): Promise<number> {
+    }
+  }
+
+  if (options.includeWorkspaceRoot === true) {
```
Don't think we can here, it's not a conflict but more of a dependent option

MalickBurger commented on src/cli.ts:
```diff
@@ -238,6 +272,23 @@ export async function run (process: NodeJS.Process): Promise<number> {
+    options.workspaces = undefined
+  } else if (options.workspaces === false) {
+    if (options.workspace.length > 0) {
+      throw new Error('Can not use --no-workspaces and --workspace at the same time')
```
Done

MalickBurger commented:
> @jkowalleck I used the branch build on my projects that used npm workspace. All works as expected. The CLI help and info is clear and I would use this version as it is.
> 
> Only small thing, on my zsh the help output from the --no-workspaces section is not aligned with the other, see screenshot.
> 
> <img alt="Screenshot 2025-03-17 at 13 26 11" width="735" src="https://private-user-images.githubusercontent.com/2203811/423468582-b58329f4-a3d9-4ad7-86f0-36f67ccbe6f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzNzE4OTgsIm5iZiI6MTc0MjM3MTU5OCwicGF0aCI6Ii8yMjAzODExLzQyMzQ2ODU4Mi1iNTgzMjlmNC1hM2Q5LTRhZDctODZmMC0zNmY2N2NjYmU2ZjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTlUMDgwNjM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmRiZmFkOGY0MjJjMDA1ZjQwY2FmMTg1N2IxZGExZGI3ODIxMjc5N2M4YWE2YWJkZGY3ZmEzMzQwNGE5NzQ0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.4AQIC29s6l8nMvY4Y5ZD7WHpDHlGbhBnO43RbtlcP-E">

I have made a change for this formatting issue and it is fixed.

jkowalleck commented:
love the result. will merge soon.

Thank you very much for your peri stance and championing this feature, @MalickBurger.
feel free to add yourself to the list of contributors : https://github.com/CycloneDX/cyclonedx-node-npm/blob/15da89b3f1c6646fd3a7e4877fdbb9546a69fe14/package.json#L39

MalickBurger commented:
> love the result. will merge soon.
> 
> Thank you very much for your persistence and championing this feature, @MalickBurger. feel free to add yourself to the list of contributors :
> 
> https://github.com/CycloneDX/cyclonedx-node-npm/blob/15da89b3f1c6646fd3a7e4877fdbb9546a69fe14/package.json#L39

https://github.com/CycloneDX/cyclonedx-node-npm/pull/1269
